ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/TopkKudaf.java

```diff
@@ -46,83 +44,53 @@
             final Schema returnType,
             final List<Schema> argumentTypes,
             final Class<T> clazz) {
-    super(functionName,
-        argIndexInValue,
-        () -> (T[]) Array.newInstance(clazz, topKSize),
-        returnType,
-        argumentTypes
+    super(functionName, argIndexInValue, () -> new ArrayList<>(),
+          returnType,
+          argumentTypes
     );
     this.topKSize = topKSize;
     this.returnType = returnType;
     this.argumentTypes = argumentTypes;
     this.clazz = clazz;
-    this.comparator = (v1, v2) -> {
-      if (v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
-
-      return Comparator.<T>reverseOrder().compare(v1, v2);
-    };
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public T[] aggregate(final T currentValue, final T[] aggregateValue) {
-    if (currentValue == null) {
-      return aggregateValue;
+  public ArrayList<T> aggregate(final T currentVal, final ArrayList<T> currentAggValList) {
+    if (currentVal == null) {
+      return currentAggValList;
     }
 
-    final T last = aggregateValue[aggregateValue.length - 1];
-    if (last != null && currentValue.compareTo(last) <= 0) {
-      return aggregateValue;
+    currentAggValList.add(currentVal);
+    Collections.sort(currentAggValList);
+    Collections.reverse(currentAggValList);
+    if (currentAggValList.size() > topKSize) {
+      currentAggValList.remove(currentAggValList.size() - 1);
     }
 
-    final int nullIndex = ArrayUtil.getNullIndex(aggregateValue);
-    if (nullIndex != -1) {
-      aggregateValue[nullIndex] = currentValue;
-      Arrays.sort(aggregateValue, comparator);
-      return aggregateValue;
-    }
-
-    aggregateValue[aggregateValue.length - 1] = currentValue;
-    Arrays.sort(aggregateValue, comparator);
-    return aggregateValue;
+    return currentAggValList;
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public Merger<String, T[]> getMerger() {
-    return (aggKey, aggOne, aggTwo) -> {
-      final T[] merged = (T[]) Array.newInstance(clazz, topKSize);
-
-      int idx1 = 0;
-      int idx2 = 0;
-      for (int i = 0; i != topKSize; ++i) {
-        final T v1 = idx1 < aggOne.length ? aggOne[idx1] : null;
-        final T v2 = idx2 < aggTwo.length ? aggTwo[idx2] : null;
-
-        if (comparator.compare(v1, v2) < 0) {
-          merged[i] = v1;
-          idx1++;
-        } else {
-          merged[i] = v2;
-          idx2++;
-        }
+  public Merger<String, ArrayList<T>> getMerger() {
+    return (aggKey, aggOneList, aggTwoList) -> {
+
+      ArrayList<T> mergedList = new ArrayList<>(aggOneList);
```

Comment:
Totally agree on the performance argument. This is because Connect uses arraylist for arrays so we switch to use the same. The reason we use this in struct is that if we have struct with array fields we would be able to use the generated struct object by connect convertor and we won't need to make any changes.
I used your suggested code but had to remove `aggregateValue.ensureCapacity(topKSize);` line since I changed aggregateValue into List.